File: frontend/src/lib/api/subcontracts.js
/**
 * Subcontracts API client — Chat 47 (Build Pack 2.8-FE-i §R3.1).
 *
 * Thin axios wrappers around the v1 subcontracts endpoints. Mirrors
 * lib/api/suppliers.js shape: shared `lib/api.js` baseURL is `/api`, so
 * each path here is prefixed `/v1/...` ourselves. All functions return
 * the response body directly (`data`).
 *
 * Endpoint contract (from backend/app/routers/subcontracts.py, verified
 * Chat 47 — origin/main, alembic head 0043_document_folders):
 *
 *   GET    /v1/subcontracts?project_id=&status=&limit=&offset=
 *                                     → { items, total }
 *   GET    /v1/subcontracts/{id}      → serialised subcontract
 *   POST   /v1/subcontracts           → 201 + serialised subcontract
 *   PATCH  /v1/subcontracts/{id}      → updated subcontract
 *   POST   /v1/subcontracts/{id}/activate   → updated subcontract
 *   POST   /v1/subcontracts/{id}/complete   → updated subcontract
 *   POST   /v1/subcontracts/{id}/terminate  → updated subcontract
 *
 * Error mapping (mirrored in hook callers + components):
 *   404 — not found / cross-tenant.
 *   409 — state error (e.g. activate non-Draft, activate unsigned,
 *         complete non-Active, terminate already-terminal). The UI
 *         MUST treat 409 distinctly from 422: surface the server
 *         `detail`, then refetch to resync the displayed status.
 *   422 — validation error (server `detail` shown).
 *   403 — missing permission.
 *
 * Sensitive contract-sum fields (`original_contract_sum`,
 * `current_contract_sum`) come back as `null` for callers without
 * `subcontracts.view_sensitive`; render via the existing em-dash
 * fallback in <SensitiveValue/> when null.
 *
 * Build-Pack §R3.1 wire contracts pinned in __tests__/subcontracts.test.js:
 *   - createSubcontract body MUST NOT include `reference` (backend
 *     generates it) or `status` (transitions go via action endpoints).
 *   - updateSubcontract body uses Pydantic `extra:"forbid"` server-side;
 *     the caller is responsible for trimming to the allowed set
 *     (`title, scope_description, original_contract_sum, retention_pct,
 *     cis_applies, start_on, end_on, signed_at, signed_by,
 *     purchase_order_id`). Sending `project_id` / `subcontractor_id` /
 *     `status` would 422.
 */
import { api } from '@/lib/api';

// ─── List + read ────────────────────────────────────────────────────

/**
 * List subcontracts visible to the caller.
 *
 * Params (all optional):
 *   - projectId — server-side filter on `project_id`.
 *   - status    — one of 'Draft' | 'Active' | 'Completed' | 'Terminated'.
 *   - limit     — 1..200 (server-clamped).
 *   - offset    — ≥0.
 *
 * Note (Build Pack §R4.1): the endpoint has NO `subcontractor_id` filter;
 * the supplier Contracts tab fetches the visible set then filters
 * client-side by `subcontractor_id === supplierId`. Surfaced in §R9
 * backlog as a future backend follow-up.
 */
export async function listSubcontracts(
  { projectId, status, limit, offset, signal } = {},
) {
  const params = {};
  if (projectId != null) params.project_id = projectId;
  if (status != null) params.status = status;
  if (limit != null) params.limit = limit;
  if (offset != null) params.offset = offset;
  const { data } = await api.get('/v1/subcontracts', { signal, params });
  return data;
}

export async function getSubcontract(subcontractId, { signal } = {}) {
  const { data } = await api.get(
    `/v1/subcontracts/${subcontractId}`, { signal },
  );
  return data;
}

// ─── Create + edit ──────────────────────────────────────────────────

/**
 * Create a Draft subcontract.
 *
 * Body shape (mirrors `SubcontractCreateBody`):
 *   { project_id, subcontractor_id, title,
 *     scope_description?, purchase_order_id?,
 *     original_contract_sum?, retention_pct?, cis_applies?,
 *     start_on?, end_on? }
 *
 * Caller MUST omit `reference` (backend generates SC-NNNN) and `status`
 * (defaults to Draft; transitions go via action endpoints).
 */
export async function createSubcontract(body) {
  const { data } = await api.post('/v1/subcontracts', body);
  return data;
}

/**
 * Patch a subcontract.
 *
 * Server uses Pydantic `extra:"forbid"`. The caller must send ONLY
 * fields in `SubcontractUpdateBody`:
 *   title, scope_description, original_contract_sum, retention_pct,
 *   cis_applies, start_on, end_on, signed_at, signed_by,
 *   purchase_order_id.
 *
 * Sending project_id / subcontractor_id / status → 422.
 */
export async function updateSubcontract(subcontractId, body) {
  const { data } = await api.patch(
    `/v1/subcontracts/${subcontractId}`, body,
  );
  return data;
}

// ─── Lifecycle transitions ──────────────────────────────────────────
//
// Action endpoints take no body. On a 409 (e.g. activating a non-Draft,
// activating an unsigned subcontract, completing a non-Active,
// terminating a terminal) the caller surfaces the server `detail` and
// refetches the detail+list to resync.

export async function activateSubcontract(subcontractId) {
  const { data } = await api.post(
    `/v1/subcontracts/${subcontractId}/activate`, {},
  );
  return data;
}

export async function completeSubcontract(subcontractId) {
  const { data } = await api.post(
    `/v1/subcontracts/${subcontractId}/complete`, {},
  );
  return data;
}

export async function terminateSubcontract(subcontractId) {
  const { data } = await api.post(
    `/v1/subcontracts/${subcontractId}/terminate`, {},
  );
  return data;
}

File: backend/app/routers/subcontracts.py
"""Subcontracts router — Chat 34 §R4 (Prompt 2.8a).

Mounted under /api/v1 via `server.py`.

Endpoints:
  POST   /subcontracts                          subcontracts.create  → 201
  GET    /subcontracts?project_id=&status=      subcontracts.view
  GET    /subcontracts/{id}                     subcontracts.view
  PATCH  /subcontracts/{id}                     subcontracts.edit
  POST   /subcontracts/{id}/activate            subcontracts.approve
  POST   /subcontracts/{id}/complete            subcontracts.edit
  POST   /subcontracts/{id}/terminate           subcontracts.approve

Error mapping:
  - SubcontractNotFoundError → 404 (cross-tenant or missing)
  - SubcontractStateError    → 409 (bad workflow transition)
  - ValueError               → 422 (payload validation)
"""
from __future__ import annotations

import uuid
from datetime import date, datetime
from typing import Any, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, Request
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.auth.deps import get_current_user, require_permission
from app.auth.permissions import UserPermissions
from app.db import get_db
from app.models.user import User
from app.services import subcontracts as svc
from app.services.subcontracts import (
    SubcontractNotFoundError, SubcontractStateError,
)


router = APIRouter(tags=["subcontracts"])


# ---------------------------------------------------------------------------
# Schemas
# ---------------------------------------------------------------------------

class SubcontractCreateBody(BaseModel):
    project_id: uuid.UUID
    subcontractor_id: uuid.UUID
    title: str = Field(..., max_length=200)
    scope_description: Optional[str] = None
    purchase_order_id: Optional[uuid.UUID] = None
    # Pack 3.5 — optional package link (NULL = standalone subcontract).
    package_id: Optional[uuid.UUID] = None
    original_contract_sum: str = "0"
    retention_pct: str = "0"
    cis_applies: bool = True
    start_on: Optional[date] = None
    end_on: Optional[date] = None


class SubcontractUpdateBody(BaseModel):
    title: Optional[str] = Field(None, max_length=200)
    scope_description: Optional[str] = None
    original_contract_sum: Optional[str] = None
    retention_pct: Optional[str] = None
    cis_applies: Optional[bool] = None
    start_on: Optional[date] = None
    end_on: Optional[date] = None
    signed_at: Optional[datetime] = None
    signed_by: Optional[uuid.UUID] = None
    purchase_order_id: Optional[uuid.UUID] = None

    # Allow explicit `null` to clear scope_description / signed_*.
    model_config = {"extra": "forbid"}


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _map(exc: Exception) -> HTTPException:
    if isinstance(exc, SubcontractNotFoundError):
        return HTTPException(status_code=404, detail=str(exc) or "Not found")
    if isinstance(exc, SubcontractStateError):
        return HTTPException(status_code=409, detail=str(exc))
    if isinstance(exc, ValueError):
        return HTTPException(status_code=422, detail=str(exc))
    return HTTPException(status_code=500, detail=str(exc))


def _serialise_with_perms(
    s, *, perms: UserPermissions,
) -> dict[str, Any]:
    has_sensitive = perms.has("subcontracts.view_sensitive")
    return svc.serialise(s, with_sensitive=has_sensitive)


# ---------------------------------------------------------------------------
# Routes
# ---------------------------------------------------------------------------

@router.post("/subcontracts", status_code=201)
def create_subcontract(
    body: SubcontractCreateBody,
    request: Request,
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(
        require_permission("subcontracts.create")
    ),
    db: Session = Depends(get_db),
):
    try:
        s = svc.create_subcontract(
            db,
            project_id=body.project_id,
            subcontractor_id=body.subcontractor_id,
            title=body.title,
            scope_description=body.scope_description,
            purchase_order_id=body.purchase_order_id,
            package_id=body.package_id,  # Pack 3.5
            original_contract_sum=body.original_contract_sum,
            retention_pct=body.retention_pct,
            cis_applies=body.cis_applies,
            start_on=body.start_on,
            end_on=body.end_on,
            user=current, perms=perms, request=request,
        )
    except (SubcontractNotFoundError, ValueError) as exc:
        raise _map(exc)
    db.commit()
    db.refresh(s)
    return _serialise_with_perms(s, perms=perms)


@router.get("/subcontracts")
def list_subcontracts(
    project_id: Optional[uuid.UUID] = Query(None),
    status: Optional[str] = Query(None),
    limit: int = Query(50, ge=1, le=200),
    offset: int = Query(0, ge=0),
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(require_permission("subcontracts.view")),
    db: Session = Depends(get_db),
):
    try:
        rows = svc.list_subcontracts(
            db, user=current, perms=perms,
            project_id=project_id, status=status,
            limit=limit, offset=offset,
        )
    except (SubcontractNotFoundError, ValueError) as exc:
        raise _map(exc)
    items = [_serialise_with_perms(r, perms=perms) for r in rows]
    return {"items": items, "total": len(items)}


@router.get("/subcontracts/{sc_id}")
def get_subcontract(
    sc_id: uuid.UUID,
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(require_permission("subcontracts.view")),
    db: Session = Depends(get_db),
):
    try:
        s = svc.get_subcontract(db, sc_id, user=current, perms=perms)
    except SubcontractNotFoundError as exc:
        raise _map(exc)
    return _serialise_with_perms(s, perms=perms)


@router.patch("/subcontracts/{sc_id}")
def update_subcontract(
    sc_id: uuid.UUID,
    body: SubcontractUpdateBody,
    request: Request,
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(require_permission("subcontracts.edit")),
    db: Session = Depends(get_db),
):
    # Only include fields the client explicitly sent so PATCH semantics hold.
    payload = body.model_dump(exclude_unset=True)
    try:
        s = svc.update_subcontract(
            db, sc_id, user=current, perms=perms,
            payload=payload, request=request,
        )
    except (SubcontractNotFoundError, SubcontractStateError, ValueError) as exc:
        raise _map(exc)
    db.commit()
    db.refresh(s)
    return _serialise_with_perms(s, perms=perms)


def _do_transition(svc_fn, *, db, sc_id, current, perms, request, **kwargs):
    try:
        s = svc_fn(
            db, sc_id, user=current, perms=perms, request=request, **kwargs,
        )
    except (SubcontractNotFoundError, SubcontractStateError, ValueError) as exc:
        raise _map(exc)
    db.commit()
    db.refresh(s)
    return _serialise_with_perms(s, perms=perms)


@router.post("/subcontracts/{sc_id}/activate")
def activate_subcontract(
    sc_id: uuid.UUID, request: Request,
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(
        require_permission("subcontracts.approve")
    ),
    db: Session = Depends(get_db),
):
    return _do_transition(
        svc.activate_subcontract,
        db=db, sc_id=sc_id, current=current, perms=perms, request=request,
    )


@router.post("/subcontracts/{sc_id}/complete")
def complete_subcontract(
    sc_id: uuid.UUID, request: Request,
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(require_permission("subcontracts.edit")),
    db: Session = Depends(get_db),
):
    return _do_transition(
        svc.complete_subcontract,
        db=db, sc_id=sc_id, current=current, perms=perms, request=request,
    )


@router.post("/subcontracts/{sc_id}/terminate")
def terminate_subcontract(
    sc_id: uuid.UUID, request: Request,
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(
        require_permission("subcontracts.approve")
    ),
    db: Session = Depends(get_db),
):
    return _do_transition(
        svc.terminate_subcontract,
        db=db, sc_id=sc_id, current=current, perms=perms, request=request,
    )
